### PR TITLE
[WIP]add the feature that can create a new engine with FlutterEngineGroup …

### PR DESCRIPTION
**Pull request by kylinchen**
add the feature that can create a new engine with FlutterEngineGroup for FlutterActivity/FlutterFragment.  

FlutterActivity/FlutterFragment adds support for multiple engines. The user sets the FlutterEngineGroup ID, calls the withNewEngineInGroup method, and creates a FlutterEngine in this FlutterEngineGroup.

fix issue:
https://github.com/flutter/flutter/issues/72742

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @blasten Do you have the cycles to review this soon?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 1]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
     }
   }
 
+  /**
+   * Creates a {@link NewEngineInGroupIntentBuilder}, which can be used to configure an {@link
+   * Intent} to launch a {@code FlutterActivity} that internally uses an existing {@link
+   * io.flutter.embedding.engine.FlutterEngineGroup} that is cached in {@link
+   * io.flutter.embedding.engine.FlutterEngineGroupCache}. and creates a new {@link
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:430]:
```suggestion
   * io.flutter.embedding.engine.FlutterEngineGroupCache}, and creates a new {@link
```

**Comment by blasten** [comment]
This is a great PR.  

First pass review, I'd like to see maybe a bit more code snippets in the code about how to use these APIs.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 2]
```diff
@@ -702,6 +705,260 @@ protected Bundle createArgs() {
...
+      this.initialRoute = initialRoute;
+      return this;
+    }
+
+    /**
+     * Whether to handle the deeplinking from the {@code Intent} automatically if the {@code
+     * getInitialRoute} returns null.
+     */
+    @NonNull
+    public NewEngineInGroupFragmentBuilder handleDeeplinking(@NonNull Boolean handleDeeplinking) {
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:799]: can this be `boolean`?

**Comment by chinmaygarde** [comment]
Is progress being made on this issue? Looks like @blasten has a few open suggestions.

**Comment by gaaclarke** [comment]
I closed out the iOS implementation saying that convenience functions can be implemented by users and incorporating them into the API doesn't scale.

**Comment by ColdPaleLight** [comment]
> Is progress being made on this issue? Looks like @blasten has a few open suggestions.

Yes, it will update soon.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 3]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
     }
   }
 
+  /**
+   * Creates a {@link NewEngineInGroupIntentBuilder}, which can be used to configure an {@link
+   * Intent} to launch a {@code FlutterActivity} that internally uses an existing {@link
+   * io.flutter.embedding.engine.FlutterEngineGroup} that is cached in {@link
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:429]: that is cached in **a** FlutterEngineGroup

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 1 above)
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:430]: to launch a FlutterActivity by internally creating a FlutterEngine from an existing FlutterEngineGroup cached in a specified FlutterEngineGroupCache

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 4]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
...
+     * @param engineGroupId The engine group id.
+     */
+    public NewEngineInGroupIntentBuilder(
+        @NonNull Class<? extends FlutterActivity> activityClass, @NonNull String engineGroupId) {
+      this.activityClass = activityClass;
+      this.cachedEngineGroupId = engineGroupId;
+    }
+
+    /**
+     * The Dart entrypoint that will be executed as soon as the Dart snapshot is loaded, default to
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:472]: The Dart entrypoint that will be executed in the newly created FlutterEngine as soon as the Dart snapshot is loaded

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 5]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
...
+     *
+     * <p>A {@code FlutterActivity} that is configured with a background mode of {@link
+     * BackgroundMode#transparent} must have a theme applied to it that includes the following
+     * property: {@code <item name="android:windowIsTranslucent">true</item>}.
+     *
+     * @param backgroundMode The background mode.
+     * @return The engine group intent builder.
+     */
+    @NonNull
+    public NewEngineInGroupIntentBuilder backgroundMode(@NonNull BackgroundMode backgroundMode) {
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:517]: rather than copying all this stuff again, can't we have them both be in the same builder class? Just have 2 constructors. One with an engine id and one with a engine group id.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 6]
```diff
@@ -214,6 +216,10 @@ void onAttach(@NonNull Context context) {
    * <p>Second, the {@code host} is given an opportunity to provide a {@link
    * io.flutter.embedding.engine.FlutterEngine} via {@link Host#provideFlutterEngine(Context)}.
    *
+   * <p>Third, the {@code host} is given an FluttenEngineGroup Id, then used a cached {@link
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:219]: this statement doesn't match the code below? The host is asked whether it would like to provide a cached FlutterEngineGroup id rather than being given one

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 7]
```diff
@@ -242,6 +248,34 @@ void onAttach(@NonNull Context context) {
...
+      FlutterEngineGroup flutterEngineGroup =
+          FlutterEngineGroupCache.getInstance().get(cachedEngineGroupId);
+      if (flutterEngineGroup == null) {
+        throw new IllegalStateException(
+            "The requested cached FlutterEngineGroup did not exist in the FlutterEngineGroupCache: '"
+                + cachedEngineGroupId
+                + "'");
+      }
+
+      String appBundlePathOverride = host.getAppBundlePath();
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:264]: this block seems unrelated to the proposed change and is a scope creep right?

File: shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroupCache.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.Map;
+
+/**
+ * Static singleton cache that holds {@link io.flutter.embedding.engine.FlutterEngineGroup}
+ * instances identified by {@code String}s.
+ *
+ * <p>The ID of a given {@link io.flutter.embedding.engine.FlutterEngineGroup} can be whatever
+ * {@code String} is desired.
+ *
+ * <p>{@code FlutterEngineGroupCache} is useful for storing pre-warmed {@link
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroupCache.java:20]: this sentence makes sense for FlutterEngine, less for FlutterEngineGroups. e.g. a FlutterEngineGroup can't exactly be "pre-warmed"

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 9]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
...
+     *
+     * <p>Choosing a background mode of {@link BackgroundMode#transparent} will configure the inner
+     * {@link FlutterView} of this {@code FlutterActivity} to be configured with a {@link
+     * FlutterTextureView} to support transparency. This choice has a non-trivial performance
+     * impact. A transparent background should only be used if it is necessary for the app design
+     * being implemented.
+     *
+     * <p>A {@code FlutterActivity} that is configured with a background mode of {@link
+     * BackgroundMode#transparent} must have a theme applied to it that includes the following
+     * property: {@code <item name="android:windowIsTranslucent">true</item>}.
```
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:511]: yes, in app/src/main/res/values/styles.xml, we can define custom theme for activity, and set the background of activity.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 10]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
...
   }
 
+  /**
+   * Creates a {@link NewEngineInGroupIntentBuilder}, which can be used to configure an {@link
+   * Intent} to launch a {@code FlutterActivity} that internally uses an existing {@link
+   * io.flutter.embedding.engine.FlutterEngineGroup} that is cached in {@link
+   * io.flutter.embedding.engine.FlutterEngineGroupCache}. and creates a new {@link
+   * io.flutter.embedding.engine.FlutterEngine} by FlutterEngineGroup#createAndRunEngine
+   *
+   * @param engineGroupId A cached engine group ID.
```
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:433]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:511]: yes, in app/src/main/res/values/styles.xml, we can define custom theme for activity, and set the background of activity.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:447]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 11]
```diff
@@ -421,6 +423,120 @@ public Intent build(@NonNull Context context) {
...
+    private final String cachedEngineGroupId;
+    private String dartEntrypoint = DEFAULT_DART_ENTRYPOINT;
+    private String initialRoute = DEFAULT_INITIAL_ROUTE;
+    private String backgroundMode = DEFAULT_BACKGROUND_MODE;
+
+    /**
+     * Constructor that allows this {@code NewEngineInGroupIntentBuilder} to be used by subclasses
+     * of {@code FlutterActivity}.
+     *
+     * <p>Subclasses of {@code FlutterActivity} should provide their own static version of {@link
```
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:469]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 12]
```diff
@@ -702,6 +705,260 @@ protected Bundle createArgs() {
     }
   }
 
+  /**
+   * Returns a {@link NewEngineInGroupFragmentBuilder} to create a {@code FlutterFragment} with a
+   * cached {@link io.flutter.embedding.engine.FlutterEngineGroup} in {@link
+   * io.flutter.embedding.engine.FlutterEngineGroupCache}. creates a new {@link
```
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 13]
```diff
@@ -702,6 +705,260 @@ protected Bundle createArgs() {
...
+   * .someExistingProperty(...) .someNewProperty(...) .build<MyFlutterFragment>(); }
+   */
+  public static class NewEngineInGroupFragmentBuilder {
+    private final Class<? extends FlutterFragment> fragmentClass;
+    private final String cachedEngineGroupId;
+    private String dartEntrypoint = "main";
+    private String initialRoute = "/";
+    private boolean handleDeeplinking = false;
+    private RenderMode renderMode = RenderMode.surface;
+    private TransparencyMode transparencyMode = TransparencyMode.transparent;
```
[the remark this answers is not in the document]
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 3 above)
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:429]: yes,  when we use this method. first, we should create a FlutterEngineGroup and cached in FlutterEngineGroupCache by flutterEngineGroupId. then we can use FlutterActivity.withNewEngineInGroup(flutterEngineGroupId) to create intent and start FlutterActivity

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 5 above)
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:548]: yes , we can  have them both be in the same builder class. But this break Single Responsibility Principle.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java (hunk 6 above)
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:219]: yes, you are right. I will update

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java (hunk 7 above)
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:264]: because the Constructor of DartEntrypoint need 2 params. One is AppBundlePath,  this block provide it from Host or FlutterLoader

File: shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroupCache.java (hunk 8 above)
Comment (kylinchen) [reply shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroupCache.java:20]: yes, you are right. I will update

**Comment by ColdPaleLight** [comment]
Friendly ping @blasten @xster

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 14]
```diff
@@ -702,6 +705,263 @@ protected Bundle createArgs() {
     }
   }
 
+  /**
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:708]: same comments on the doc for from FlutterActivity on the next 2 classes

File: shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroup.java [hunk 15]
```diff
@@ -178,7 +178,7 @@ public void onEngineWillDestroy() {
   }
 
   @VisibleForTesting
-  /* package */ FlutterEngine createEngine(Context context) {
+  public FlutterEngine createEngine(Context context) {
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroup.java:181]: you don't need to do this right?

**Review by xster — COMMENTED** [review]
sorry for the wait. I was away last month. Almost there.

**Comment by chinmaygarde** [comment]
@kylinchen Any updates after @xster s last review?

**Comment by chinmaygarde** [comment]
Closing as this seems stale. Please re-open once progress can be made. Thanks.

**Comment by ColdPaleLight** [comment]
Talked to @kylinchen offline and he will restart work on this PR recently. So I reopened the PR and marked it as "WIP"

**Comment by chinmaygarde** [comment]
Are any updates on this likely?

**Comment by ColdPaleLight** [comment]
I'll take over this PR and close it for now. I will reopen it in the future.
